### PR TITLE
Support SUNDIALS 4.1.0

### LINKED ISSUE #913: Configure checks not strict enough for SUNDIALS
We need `nvector_parallel.h`, which is not in the version of SUNDIALS available on Ubuntu 14 (used on Travis), but the configure checks pass.

This is because we check for `sundials-config`, query it for what flags to use and use them -- we don't check it has parallel capabilities. I'm not sure if this is possible, as `sundials-config` as removed in v2.6.0 when they moved to CMake. I've not established if some other config checker mechanism made it back in -- there doesn't appear to be anything in v2.7.0 though.

One option would be to try to compile a test program that uses everything we need, and turn off support if that doesn't work.

On a side-note, the whole configure section for MUMPS/CVODE/IDA/ARKODE/SUNDIALS needs a good tidy.

### PR DESCRIPTION
**Pull request by ZedThree**
We now support SUNDIALS 2.6 to 4.1.0!

Fixes #913 #1644 

I've also taken the time to tidy up the three SUNDIALS solvers.

- Support SUNDIALS 3.x
- Support SUNDIALS 4.x
- Declare lots of local variables `const` and move them to either smaller scopes or to immediately before where they are actually used
  - This is mostly thanks to the new `Options` interface that allows this!
- Remove a whole bunch of `TRACE` macros as they had the same information as the exception thrown in the same scope
- Fix some small memory leaks
  - Mostly not a problem as the solver hangs around for most of the life of the program anyway
- Update SUNDIALS installation instructions
- Clang-format the SUNDIALS solvers
- Build and test with SUNDIALS on Travis
  - Compilation is (hopefully!) cached so we won't need to pay for it every time

---

I've used a slightly unusual device in this PR: [IIFE -- "immediately invoked function expression"](https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html)

Basically, we would like to always initialise variables at the same time as declaring them, as well as declaring them `const`. If the value depends on a single condition, we can use a ternary. For more complex logic like multiple conditions, we can wrap the logic up into a lambda and then immediately call it:
```cpp
  const auto& explicit_rhs = [imex, solve_explicit]() {
    if (imex) {
      return arkode_rhs_explicit;
    } else {
      return solve_explicit ? arkode_rhs : nullptr;
    }
  }();
```
They're a little bit weird at first, but once you've seen a couple they're ok. They look just like initialising a variable to some expression, with some extra punctuation.

### CODE REVIEW

**Comment by ZedThree** [comment]
I'm just attempting to build and cache SUNDIALS on Travis in another branch -- WIP while I wait for that

File: src/solver/impls/cvode/cvode.cxx [hunk 1]
```diff
@@ -28,53 +28,97 @@
...
 #include <sundials/sundials_types.h>
-#include <sundials/sundials_math.h>
 
-#include <output.hxx>
+#include <algorithm>
+#include <numeric>
+#include <string>
 
-#include "unused.hxx"
+class Field2D;
```
Comment (johnomotani) [inline src/solver/impls/cvode/cvode.cxx]: forward declaration again?

**Comment by johnomotani** [comment]
It would be nice to have some integrated tests that use ARKODE or IDA and use their features (split explicit/implicit operators or constraints). That would help to ensure that updating the libraries doesn't mess up performance on a known working case (e.g. because of changes to the way options are set). There are some models in `examples/IMEX`, but I think they might need some work to turn into tests; e.g. expected output, checking that the required solver is available...

File: manual/sphinx/user_docs/advanced_install.rst [hunk 2]
```diff
@@ -326,52 +326,45 @@ solver. Currently, BOUT++ also supports the SUNDIALS solvers CVODE, IDA
...
-     $ cd build
+     $ # Move the downloaded sundials-4.1.0.tar.gz to sundials-install
+     $ tar -xzvf sundials-4.1.0.tar.gz
+     $ mkdir build && cd build
 
      $ cmake \
        -DCMAKE_INSTALL_PREFIX=$HOME/local \
-       -DEXAMPLES_INSTALL_PATH=$HOME/local/examples \
-       -DCMAKE_LINKER=$HOME/local/lib \
        -DLAPACK_ENABLE=ON \
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply manual/sphinx/user_docs/advanced_install.rst]: Good point. It also looks like you need to explicitly use the LAPACK solvers at compile-time, which we definitely don't.

File: src/solver/impls/arkode/arkode.cxx [hunk 3]
```diff
@@ -29,54 +29,148 @@
...
-#include <sundials/sundials_types.h>
 #include <sundials/sundials_math.h>
+#include <sundials/sundials_types.h>
 
-#include <output.hxx>
+#include <algorithm>
+#include <numeric>
 
-#include "unused.hxx"
+class Field2D;
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/solver/impls/arkode/arkode.cxx]: Yes, it is a bit odd. I was just going off include-what-you-use, but I'm not sure why it thought it was necessary. I'll check again, it usually lists why you need to include headers at least

**Comment by ZedThree** [comment]
Yeah, all three SUNDIALS solvers have multiple options that affect exactly what they do, so it would be worth checking all combinations work. I'll see what I can do
